dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -432,79 +436,154 @@ SegmentPtr Segment::restoreSegment( //
 
 Segment::SegmentMetaInfos Segment::readAllSegmentsMetaInfoInRange( //
     DMContext & context,
+    std::shared_ptr<GeneralCancelHandle> cancel_handle,
     const RowKeyRange & target_range,
     const CheckpointInfoPtr & checkpoint_info)
 {
-    auto fap_context = context.global_context.getSharedContextDisagg()->fap_context;
-
+    auto log = DB::Logger::get();
+    Stopwatch sw;
+    SCOPE_EXIT(
+        { GET_METRIC(tiflash_fap_task_duration_seconds, type_write_stage_read_segment).Observe(sw.elapsedSeconds()); });
+    static constexpr UInt64 WAIT_TIME_THRESHOLD = 20;
+    // We have a cache that records all segments which map to a certain table identified by (keyspace_id, physical_table_id).
+    // We can thus avoid reading from the very beginning for every different regions in this table.
     // If cache is empty, we read from DELTA_MERGE_FIRST_SEGMENT_ID to the end and build the cache.
     // Otherwise, we just read the segment that cover the range.
     PageIdU64 current_segment_id = DELTA_MERGE_FIRST_SEGMENT_ID;
     auto end_to_segment_id_cache = checkpoint_info->checkpoint_data_holder->getEndToSegmentIdCache(
         KeyspaceTableID{context.keyspace_id, context.physical_table_id});
-    auto lock = end_to_segment_id_cache->lock();
-    bool is_cache_ready = end_to_segment_id_cache->isReady(lock);
-    if (is_cache_ready)
-    {
-        current_segment_id
-            = end_to_segment_id_cache->getSegmentIdContainingKey(lock, target_range.getStart().toRowKeyValue());
+    bool is_cache_ready = false;
+    // If there is a table building cache, then other table may block to read the built cache.
+    // If the remote reader causes much time to retrieve data, then these tasks could block here.
+    // However, when the execlusive holder is canceled due to timeout, the readers could eventually get the lock.
+    FAIL_POINT_PAUSE(FailPoints::pause_when_building_fap_segments);
+    {
+        auto sec = sw.elapsedSecondsFromLastTime();
+        // Lock acquires when:
+        // 1. No writer.
+        // 2. The writer finishes.
+        // 3. The writer is canceled.
+        auto lock = end_to_segment_id_cache->readLock();
+        is_cache_ready = end_to_segment_id_cache->isReady(lock);
+        auto el = sw.elapsedSecondsFromLastTime() - sec;
```

Comment:
```suggestion
        // Lock acquires when:
        // 1. No writer.
        // 2. The writer finishes.
        // 3. The writer is canceled.
        auto lock = end_to_segment_id_cache->readLock();
        is_cache_ready = end_to_segment_id_cache->isReady(lock);
        auto el = sw.elapsedSecondsFromLastTime();
```

Check the logic of `elapsedSecondsFromLastTime`

https://github.com/pingcap/tiflash/blob/9e3ba660eee20253548d26fdf5bb9f2fa5731e4e/dbms/src/Common/Stopwatch.h#L92-L105